Various fixes

sahilgupta opened the pull request.

sahilgupta committed:
Style fixes

sahilgupta committed:
Cache the auth token

Obtain the auth token only once, at the time of initialization and cache it so that it gets re-used on subsequent requests.

ayushgoel commented on mstranslator/translator.py:
```diff
@@ -14,55 +18,67 @@ def __init__(self, translator_client_id, translator_client_secret):
     def __init__(self, config):
-        self.config = config
+        assert isinstance(config, Config) is True
+        self.access_token = self._get_access_token(config)
```
`access_token` can not be cached infinitely (which can be the lifetime of the object). You will need to recreate `access_token` by parsing `expires_in` attribute of the `access_token` call.
Read: https://msdn.microsoft.com/en-us/library/hh454950.aspx

ayushgoel commented on mstranslator/translator.py:
```diff
@@ -1,9 +1,13 @@
-import requests
+"""Python wrapper for Microsoft Translate services."""
```
Is this documentation useful anywhere? Should it not be instead put up in [setup?](https://github.com/ayushgoel/mstranslator/blob/master/setup.py#L7)

sahilgupta replied:
http://stackoverflow.com/a/2557196

ayushgoel commented on mstranslator/translator.py:
```diff
@@ -1,9 +1,13 @@
+
 
-class Config:
+class Config(object):
```
I was unable to find a reason to inherit from object. Please let me know the reason.

sugandha7 replied:
https://docs.python.org/release/2.2.3/whatsnew/sect-rellinks.html

ayushgoel replied:
Thanks @sugandha7 ! :+1:

ayushgoel commented on mstranslator/translator.py:
```diff
@@ -1,9 +1,13 @@
     """Config to be given to an instance of translator to do the Authorization."""
+
     def __init__(self, translator_client_id, translator_client_secret):
         assert translator_client_id is not None
```
Now you need to call `super.__init__`.

sahilgupta replied:
Na, no need. Call to `object`'s `__init__` doesn't do anything.

There's no behaviour of the parent that we need to run.

ayushgoel replied:
The super class can change. Methods should follow the inheritance chain without considering what super class implementation does.

sahilgupta replied:
But you might not even want parent class's behaviour? Why not implement only when there's actual need?

ayushgoel replied:
> But you might not even want parent class's behaviour? 
> Why not implement only when there's actual need?

Then why inherit at all?
Why inherit now?

sahilgupta replied:
Well, for one, new style classes fix some issues which might get faced by someone inheriting _from_ this class.
Inheriting from object is not about what this class does, but what one further down the chain might want to do.

Secondly, old style classes don't work with Python3.

ayushgoel replied:
1. Are we debating, "Should we call super's implementation on an overridden method?"?
2. Or is it, "Classes should inherit from `object`?"

Because for 2. we have agreed that subclassing the way to go.

sahilgupta replied:
The latter, but don't consider the entire thing even remotely significant enough to spend more discussing.
Add a `super`call, that's it?

You can do it while merging if everything else is good to go.

ayushgoel replied:
> Add a supercall, that's it?

Yes, that was it. Since you are not willing to do that, I will (have to) do it. 
Thank you for your contribution. :+1:

ayushgoel commented on mstranslator/translator.py:
```diff
@@ -14,55 +18,67 @@ def __init__(self, translator_client_id, translator_client_secret):
+
     def __init__(self, config):
-        self.config = config
+        assert isinstance(config, Config) is True
```
is `type` better than `isinstance`? 
Check: https://github.com/ayushgoel/mstranslator/pull/1/files#diff-6b603b02745072005b6a815bb2954d72R13

sahilgupta replied:
Sort of. `isinstance` caters for inheritance, while checking for equality of `type` does not.

Check http://stackoverflow.com/a/1549854 for more.

ayushgoel commented on mstranslator/translator.py:
```diff
@@ -14,55 +18,67 @@ def __init__(self, translator_client_id, translator_client_secret):
+            "scope": 'http://api.microsofttranslator.com',
+            "grant_type": 'client_credentials'
+        }
+        resp = requests.post(url='https://datamarket.accesscontrol.windows.net/v2/OAuth2-13',
```
Should the `=` be wrapped with whitespace? I am seeing that being done in PEP-8 http://legacy.python.org/dev/peps/pep-0008/.

sahilgupta replied:
almost every operator should be wrapped with spaces.

ayushgoel commented on mstranslator/translator.py:
```diff
@@ -14,55 +18,67 @@ def __init__(self, translator_client_id, translator_client_secret):
+        }
+        resp = requests.post(url='https://datamarket.accesscontrol.windows.net/v2/OAuth2-13',
+                             data=urllib.urlencode(data))
         return resp.json()["access_token"]
```
Check `resp.ok`?

ayushgoel commented on mstranslator/translator.py:
```diff
@@ -14,55 +18,67 @@ def __init__(self, translator_client_id, translator_client_secret):
-            raise
+        resp = requests.get(url='http://api.microsofttranslator.com/v2/Http.svc/Detect',
+                            params=data, headers=headers)
+        if resp.ok:
```
:clap: :bow:

ayushgoel commented on mstranslator/translator.py:
```diff
@@ -14,55 +18,67 @@ def __init__(self, translator_client_id, translator_client_secret):
-    def __authorization_header(self):
-        access_token = self.__get_access_token()
-        return "Bearer" + " " + access_token
+    def _authorization_header(self):
```
I thought functions you intend to keep private should be prefixed with `_`. Not true?

sahilgupta replied:
prefixed with a single `_` not double `__`

ayushgoel replied:
:question:

ayushgoel commented on mstranslator/translator.py:
```diff
@@ -14,55 +18,67 @@ def __init__(self, translator_client_id, translator_client_secret):
+            raise resp.raise_for_status()
 
-    def translate(self, text, from_language, to_language, authorization_header):
-        if authorization_header is None:
```
This will need review once [this](https://github.com/ayushgoel/mstranslator/pull/1/files#r56211710) is fixed.

sahilgupta committed:
Fix auth-token caching.

Auth tokens are only valid for a limited number of seconds.
Token expiry is sent by the server at the time of obtaining the token and can be used to
auto-refresh the token before making a new request, post expiry time.

sahilgupta committed:
Minor refactor

sahilgupta committed:
Lazily fetch the auth-token.

ayushgoel commented on mstranslator/translator.py:
```diff
@@ -14,55 +19,77 @@ def __init__(self, translator_client_id, translator_client_secret):
+            "scope": 'http://api.microsofttranslator.com',
+            "grant_type": 'client_credentials'
+        }
+        resp = requests.post(url='https://datamarket.accesscontrol.windows.net/v2/OAuth2-13',
```
Wrap `=` with spaces.

sahilgupta replied:
Oh, no spaces go around keyword arguments.

ayushgoel commented on mstranslator/translator.py:
```diff
@@ -14,55 +19,77 @@ def __init__(self, translator_client_id, translator_client_secret):
     def __init__(self, config):
+        assert isinstance(config, Config) is True
         self.config = config
+        self.access_token, self.token_expiry = self._get_access_token()
```
Why do this in an initializer? Let it be lazy and be generated when required.

sahilgupta replied:
Was debating whether to do that or not. Fixed.

ayushgoel commented on mstranslator/translator.py:
```diff
@@ -14,55 +19,77 @@ def __init__(self, translator_client_id, translator_client_secret):
-        resp = requests.post(url='https://datamarket.accesscontrol.windows.net/v2/OAuth2-13', data=urllib.urlencode(data))
-        return resp.json()["access_token"]
+        token_expiry = time.time() + int(resp.json()["expires_in"])
+        access_token = resp.json()["access_token"]
```
Extract `resp.json()` to a var.

ayushgoel commented on mstranslator/translator.py:
```diff
@@ -14,55 +19,77 @@ def __init__(self, translator_client_id, translator_client_secret):
+
+    def _authorization_header(self):
+        if (not self.token_expiry) or (self.token_expiry < time.time()):
+            self.access_token, self.token_expiry = self._get_access_token()
```
I didn't want to introduce access_token caching. Because it is just hard. We might need to add tests to have some faith on this. :sob: 
Also, :clap: for giving this a shot.
